projects/plugins/jetpack/_inc/client/main.jsx

```diff
@@ -208,6 +210,33 @@ class Main extends React.Component {
 	}
 
 	renderMainContent = route => {
+		/*
+		 * Show "Partner Coupon Redeem" screen instead of regular main content/pre-connection.
+		 */
+		if ( this.props.partnerCoupon ) {
+			const forceShow = new URLSearchParams( window.location.search ).get( 'showCouponRedemption' );
+
+			/*
+			 * There are two conditions (groups of conditions, really) where we would want to
+			 * show the partner coupon redeem screen:
+			 *
+			 * 1. The site is not yet connected to WPCOM, but has the jetpack_partner_coupon
+			 * option set in the database (this.props.partnerCoupon in redux). This is likely a
+			 * partner-user who has just arrived here from a CTA within a partner's dashboard
+			 * or other ecosystem.
+			 *
+			 * 2. The site is already connected to WPCOM, but the jetpack_partner_coupon option
+			 * is still set in the database. This means the user connected their site, but never
+			 * redeemed the coupon. If this is the case, we don't want to override the dashboard
+			 * or at a glance pages with the redemption screen. Instead, we'll catch a URL
+			 * parameter that JITMs will set (showCouponRedemption=true), and show the screen only
+			 * when the user came from a a JITM.
+			 */
+			if ( ! this.props.isSiteConnected || forceShow ) {
```

Comment:
Should the site connection take priority, so if the site is connected, we do not display the banner even if you use the query string?

Reply:
It's on purpose that we've added a query string override. It's so we can link directly to the coupon component in "remote JITMs" where we're going to do some coupon validation on the server-side (e.g.: has the coupon already been redeemed?).
It might be that you have a use-case in mind that we've missed but it should be OK 😄